src/dotnet/MLSearch.Service/Bot/Tools/Context/BotToolsContextHandler.cs

```diff
@@ -0,0 +1,95 @@
+using System.IdentityModel.Tokens.Jwt;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.IdentityModel.Tokens;
+
+namespace ActualChat.MLSearch.Bot.Tools.Context;
+public class BotToolsContextHandlerOptions {
+    public SigningCredentials SigningCredentials{ get; set; }
+    public string Audience { get; set; }
+    public string Issuer { get; set; }
+    public TimeSpan ContextLifetime { get; set; }
+}
+
+public class BotToolsContext(ClaimsPrincipal? claims): IBotToolsContext {
+    public const string ConversationClaimType = "ConversationId";
+    public string? ConversationId => claims?.FindFirstValue(ConversationClaimType);
+    public bool IsValid => claims != null 
+        && claims.HasClaim(e => string.Equals(e.Type, ConversationClaimType, StringComparison.Ordinal));
+    
+}
+
+public class BotToolsContextHandler(IOptionsMonitor<BotToolsContextHandlerOptions> options): IBotToolsContextHandler
+{
+    private const string BEARER_PREFIX = "Bearer ";
+
+    public IBotToolsContext GetContext(HttpRequest request) => new BotToolsContext(GetValidatedClaims(request));
+    public void SetContext(HttpRequestMessage request, string conversationId) {
+        var config = options.CurrentValue;
+        var claims = new List<Claim>();
+        claims.Add(new Claim(BotToolsContext.ConversationClaimType, conversationId));
+        var expiration = DateTime.Now.Add(config.ContextLifetime);
```

Comment:
It is better not to use DateTime.Now but instead use some clock (e.g. MomentClock) provided as constructor parameter.
Also, what about timezone? Now takes time in the current server timezone. If token receiver uses other timezone we may have some unexpected behavior when token lives longer than expected or expires sooner.

Reply:
- The DateTime expiration argument is always converted into NumericDate format in the JwtSecurityToken internally. (seconds since epoch).
- Since it is seconds there's no need to use MomentClock or any other precise clocks.
- It can't be a parameter in the constructor as it should use current time.

Reply:
- As datetime doesn't store timezone information internal tick count of `DateTime.Now` value depends on server timezone. Literally it offsets UtcNow to server timezone (see https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L1520).
So `JwtSecurityToken` has to internally offset time back to Utc (See https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs#L274 and https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/src/Microsoft.IdentityModel.Tokens/DateTimeUtil.cs#L85). While it works those back-and-forth conversion are avoidable should we use `DateTime.UtcNow`.
- I mentioned `MomentClock` passed via constructor parameter because it makes code testable as you are not stick to real time and can control time flow via the clock mock. This allows to get rid of flucky tests.

Reply:
I agree that DateTime.UtcNow is better. I'll change into that.
I generally do not disagree with the proposal to add a source of DateTime to a constructor where it make sense. However this is not the case here. Let me try to explain my thoughts:
- This class has two sides: one generates a token, and the second one is checking its validity.
- If we pass a source of a DateTime anywhere (a constructor / a parameter) it must be used by both methods.
- It's very easy to set an expiration time for the token, however the validation counterpart is not that easy. I've double-checked it and they use DateTime.UtcNow in it's inner static method: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/eb4df8fc5a9a4fbf50de2dd5bfd263a225c7c87a/src/Microsoft.IdentityModel.Tokens/ValidatorUtilities.cs#L39